vformer/attention/cross.py

```diff
@@ -73,3 +72,57 @@ def forward(self, cls, patches):
         ycls = self.gl(ycls)
 
         return ycls
+
+
+@ATTENTION_REGISTRY.register()
+class CrossAttention(nn.Module):
+    """
+    Cross-Attention
+
+    Parameters
+    ----------
+    query_dim: int
+        Dimension of query array
+    context_dim: int
+        Dimension of context array
+    num_heads: int
+        Number of cross-attention heads
+    head_dim: int
+        Dimension of each head
+
+    """
+
+    def __init__(self, query_dim, context_dim, num_heads=8, head_dim=64):
+        super().__init__()
+
+        inner_dim = num_heads * head_dim
+        self.num_heads = num_heads
+        self.scale = head_dim ** -0.5
+        self.to_q = nn.Linear(query_dim, inner_dim)
+        self.to_k = nn.Linear(context_dim, inner_dim)
+        self.to_v = nn.Linear(context_dim, inner_dim)
+        self.to_out = _Projection(inner_dim, query_dim)
+        self.attend = nn.Softmax(dim=-1)
+
+    def forward(self, x, context, mask=None):
+        h = self.num_heads
+        q = self.to_q(x)
+        k = self.to_k(context)
+        v = self.to_v(context)
+        k = rearrange(k, "b n (h d) -> b h n d", h=h)
+        q = rearrange(q, "b n (h d) -> b h n d", h=h)
+        v = rearrange(v, "b n (h d) -> b h n d", h=h)
+
+        attention = torch.matmul(q, k.transpose(-1, -2)) * self.scale
+
+        if mask is not None:
```

Comment:
I am not sure if we need mask, I think mask is mainly for NLP models?

Reply:
I think it would be helpful to keep the mask option. Some training strategies for vision transformers do patch level supervision in which having the ability to mask regions may be helpful. For eg, [this paper](https://arxiv.org/abs/2111.06377).